libbpf-tools/vfsstat.bpf.c

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Anton Protopopov
+//
+// Based on vfsstat(8) from BCC by Brendan Gregg
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "vfsstat.h"
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, S_MAXSTAT);
+	__type(key, int);
+	__type(value, u64);
+} stats SEC(".maps");
+
+static __always_inline int inc_stats(int key)
+{
+	u64 *val;
+
+	val = bpf_map_lookup_elem(&stats, &key);
+	if (val)
+		++*val;
+	return 0;
+}
+
```

Comment:
This is actually racy when multiple CPUs simultaneously do same vfs operation type. I'd suggest using atomic increments. Together with above suggestion to use plain array, it would look like this:

```
__atomic_add_fetch(&stats[key], 1);
```